In [1]:
import pandas as pd
import wrangle
import anomaly_detection

import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
df = pd.read_csv('anonymized-curriculum-access.txt', sep=' ', header=None)

In [3]:
df.head(2)

0         1        2  3    4             5
0  2018-01-26  09:55:03        /  1  8.0  97.105.19.61
1  2018-01-26  09:56:02  java-ii  1  8.0  97.105.19.61

In [4]:
df = df.rename(columns={0:'date',1:'time',2:'page',3:'user', 4:'cohort',5:'ip' })

In [5]:
df.head()

date      time                                 page  user  cohort  \
0  2018-01-26  09:55:03                                    /     1     8.0   
1  2018-01-26  09:56:02                              java-ii     1     8.0   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming     1     8.0   
3  2018-01-26  09:56:06   slides/object_oriented_programming     1     8.0   
4  2018-01-26  09:56:24            javascript-i/conditionals     2    22.0   

             ip  
0  97.105.19.61  
1  97.105.19.61  
2  97.105.19.61  
3  97.105.19.61  
4  97.105.19.61

In [6]:
# this is the code that begins to awser question #1
df.page.value_counts().head(10)

page
/                           50313
search/search_index.json    19519
javascript-i                18983
toc                         18297
java-iii                    13733
html-css                    13635
java-ii                     12685
spring                      12524
jquery                      11525
mysql                       11033
Name: count, dtype: int64

In [7]:
df

date      time  \
0       2018-01-26  09:55:03   
1       2018-01-26  09:56:02   
2       2018-01-26  09:56:05   
3       2018-01-26  09:56:06   
4       2018-01-26  09:56:24   
...            ...       ...   
900218  2021-04-21  16:41:51   
900219  2021-04-21  16:42:02   
900220  2021-04-21  16:42:09   
900221  2021-04-21  16:44:37   
900222  2021-04-21  16:44:39   

                                                    page  user  cohort  \
0                                                      /     1     8.0   
1                                                java-ii     1     8.0   
2                    java-ii/object-oriented-programming     1     8.0   
3                     slides/object_oriented_programming     1     8.0   
4                              javascript-i/conditionals     2    22.0   
...                                                  ...   ...     ...   
900218                              jquery/personal-site    64    28.0   
900219                                 jquery/mapbox-api    64    28.0   
900220                           jquery/ajax/weather-map    64    28.0   
900221  anomaly-detection/discrete-probabilistic-methods   744    28.0   
900222                                 jquery/mapbox-api    64    28.0   

                   ip  
0        97.105.19.61  
1        97.105.19.61  
2        97.105.19.61  
3        97.105.19.61  
4        97.105.19.61  
...               ...  
900218  71.150.217.33  
900219  71.150.217.33  
900220  71.150.217.33  
900221  24.160.137.86  
900222  71.150.217.33  

[900223 rows x 6 columns]

In [8]:
# this code gets the awnser to number one entirely
# size allows us to get the counts for two 
q_one = df.groupby(['cohort', 'page']).size().reset_index(name='count')

In [9]:
q_one

cohort                                         page  count
0         1.0                                            /    626
1         1.0                  0_Classification_Algorithms      1
2         1.0                              2.0_Intro_Stats      1
3         1.0                         5.0_Intro_Regression      2
4         1.0                         7.0_Intro_Clustering      2
...       ...                                          ...    ...
13560   139.0                                       jquery      4
13561   139.0  jquery/essential-methods/attributes-and-css      1
13562   139.0                     search/search_index.json      9
13563   139.0                                       spring      2
13564   139.0                                          toc    101

[13565 rows x 3 columns]

In [10]:
# can use basketball deep learnign code to break the cohorts apart and comare their top 5 taht way
q_one.sort_values(by='count', ascending=False).head(20)

cohort                      page  count
5922     28.0                         /   6340
11247    59.0                         /   2980
4528     24.0                         /   2101
9731     53.0                         /   2038
8643     34.0                         /   1967
12221    62.0                       toc   1866
3576     22.0                         /   1828
7038     28.0              javascript-i   1817
7325     29.0                         /   1798
10680    57.0                         /   1727
9991     55.0                         /   1712
11941    62.0                         /   1696
8342     33.0                         /   1653
4120     23.0                         /   1641
10633    56.0  search/search_index.json   1497
9176     51.0                         /   1467
7635     29.0                       toc   1465
7220     28.0                    spring   1403
7024     28.0                  java-iii   1393
8609     33.0  search/search_index.json   1380

## 4. Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

In [11]:
df.head()

date      time                                 page  user  cohort  \
0  2018-01-26  09:55:03                                    /     1     8.0   
1  2018-01-26  09:56:02                              java-ii     1     8.0   
2  2018-01-26  09:56:05  java-ii/object-oriented-programming     1     8.0   
3  2018-01-26  09:56:06   slides/object_oriented_programming     1     8.0   
4  2018-01-26  09:56:24            javascript-i/conditionals     2    22.0   

             ip  
0  97.105.19.61  
1  97.105.19.61  
2  97.105.19.61  
3  97.105.19.61  
4  97.105.19.61

In [12]:
# checking most used ip address
q_four = df.groupby(['ip']).size().reset_index(name='count')

In [13]:
q_four.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5531 entries, 0 to 5530
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   ip      5531 non-null   object
 1   count   5531 non-null   int64 
dtypes: int64(1), object(1)
memory usage: 86.5+ KB


In [14]:
q_four.head()

ip  count
0   100.11.170.135    114
1     100.19.73.35    408
2    100.35.67.162     11
3   103.137.12.164      9
4  103.208.220.130      1

In [15]:
# checking visually to see if there is a big gap between access numbers
q_four.sort_values(by='count', ascending=False).head(20)

ip   count
5097     97.105.19.58  284579
5099     97.105.19.61   61662
2827  192.171.117.210    9515
4285    71.150.217.33    6791
4777   76.185.145.231    4754
875    12.106.208.194    4360
3814      67.11.50.23    4181
3981   70.112.179.142    4117
4099    70.121.183.95    3843
1336    172.124.67.93    3511
5518    99.76.233.212    3491
3191    24.243.67.104    3232
3721     67.10.174.56    3115
4570    72.190.174.52    2939
13      104.10.49.137    2711
3070    216.63.31.195    2655
4771   76.185.131.226    2645
3588    66.196.21.182    2612
2179  173.173.115.201    2599
1046    136.50.32.151    2570

In [16]:
inner_fence, outer_fence = anomaly_detection.get_lower_and_upper_bounds(q_four['count'])

In [17]:
inner_fence

inner_fence
0             114
1             408
13           2711
25            450
31            207
...           ...
5520          403
5522          116
5523         1795
5528          267
5530          391

[928 rows x 1 columns]

In [18]:
# extreme outliers of ip access
outer_fence.sort_values(by='outer_fence')

outer_fence
4874          120
2438          120
4784          120
5107          121
4863          121
...           ...
4777         4754
4285         6791
2827         9515
5099        61662
5097       284579

[738 rows x 1 columns]

In [19]:
# there are the most used two ips that fall well outside of 
# the upper outer fence
big_out = q_four.sort_values(by='count', ascending=False).head(2)

In [30]:
big_out

ip   count
5097  97.105.19.58  284579
5099  97.105.19.61   61662

In [20]:
q_four

ip  count
0      100.11.170.135    114
1        100.19.73.35    408
2       100.35.67.162     11
3      103.137.12.164      9
4     103.208.220.130      1
...               ...    ...
5526     99.97.210.12     18
5527    99.98.245.245      2
5528     99.98.32.165    267
5529      99.98.32.97     46
5530     99.98.36.168    391

[5531 rows x 2 columns]

In [21]:
#drops teh highest two
lower_avg = q_four.sort_values(by='count', ascending=False).iloc[2:]

# this will be the average of the remaiing 
lower_avg = lower_avg['count'].mean().round()

In [22]:
lower_avg

100.0

In [23]:
# this shows visually the difference between the average of all others and our
# two suspects

## Nisha will finish below visualization
# lower_avg = df.drop(df[df['count'] == 'value'].index, inplace=True)

KeyError: 'count'

In [24]:
# checkign to see if there is anything suspect about our two big guys
# this makes the bigggest ip logs into it's own df
biggest = df[df['ip'] == '97.105.19.58']

In [25]:
biggest.head()

date      time                             page  user  cohort  \
86014  2018-07-16  07:51:57           javascript-ii/promises   211     1.0   
86015  2018-07-16  07:58:03  javascript-ii/map-filter-reduce   211     1.0   
86016  2018-07-16  07:59:50           javascript-ii/promises   211     1.0   
86017  2018-07-16  08:07:37                       index.html    53    14.0   
86018  2018-07-16  08:07:41                    javascript-ii    53    14.0   

                 ip  
86014  97.105.19.58  
86015  97.105.19.58  
86016  97.105.19.58  
86017  97.105.19.58  
86018  97.105.19.58

In [26]:
# combining date and time then making them a datetime object. 
# and index

biggest['datetime'] = pd.to_datetime(biggest['date'] + ' ' + biggest['time'])

/var/folders/__/8b23nn817hq5cnv1011jx8kr0000gn/T/ipykernel_4693/3340105441.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  biggest['datetime'] = pd.to_datetime(biggest['date'] + ' ' + biggest['time'])


In [28]:
biggest.datetime

86014    2018-07-16 07:51:57
86015    2018-07-16 07:58:03
86016    2018-07-16 07:59:50
86017    2018-07-16 08:07:37
86018    2018-07-16 08:07:41
                 ...        
491216   2020-03-23 15:10:09
491218   2020-03-23 15:10:18
491221   2020-03-23 15:20:39
491226   2020-03-23 15:24:54
491279   2020-03-23 16:07:31
Name: datetime, Length: 284579, dtype: datetime64[ns]

In [29]:
# Calculate time differences between logs
time_diffs = biggest['datetime'].diff()

# Average time between logs
avg_time_diff = time_diffs.mean()

# Shortest amount of time between logs
min_time_diff = time_diffs.min()

# Frequency of the shortest time difference
min_time_diff_freq = time_diffs.value_counts().get(min_time_diff)

print(f"Average time between logs: {avg_time_diff}")
print(f"Shortest amount of time between logs: {min_time_diff}")
print(f"Frequency of the shortest time difference: {min_time_diff_freq}")


Average time between logs: 0 days 00:03:07.126671773
Shortest amount of time between logs: 0 days 00:00:00
Frequency of the shortest time difference: 12310


In [33]:
biggest.head()

date      time                             page  user  cohort  \
86014  2018-07-16  07:51:57           javascript-ii/promises   211     1.0   
86015  2018-07-16  07:58:03  javascript-ii/map-filter-reduce   211     1.0   
86016  2018-07-16  07:59:50           javascript-ii/promises   211     1.0   
86017  2018-07-16  08:07:37                       index.html    53    14.0   
86018  2018-07-16  08:07:41                    javascript-ii    53    14.0   

                 ip            datetime  
86014  97.105.19.58 2018-07-16 07:51:57  
86015  97.105.19.58 2018-07-16 07:58:03  
86016  97.105.19.58 2018-07-16 07:59:50  
86017  97.105.19.58 2018-07-16 08:07:37  
86018  97.105.19.58 2018-07-16 08:07:41

In [40]:
biggest.time.min()

'06:09:34'

In [41]:
biggest.time.max()

'23:34:11'

In [42]:
biggest.user.value_counts()

user
11     7049
64     5690
53     4895
314    3474
1      3099
       ... 
619       1
629       1
461       1
131       1
621       1
Name: count, Length: 402, dtype: int64

### 97.105.19.58 
### This IP is shady as all hell. 
Average time between logs: 0 days 00:03:07.126671773
Shortest amount of time between logs: 0 days 00:00:00
Frequency of the shortest time difference: 12310
Does not seem like a human operator

It is associated with many cohorts the earliest time it pops up at is 06:09 but the latest is just before midnitght 23:34:11.

402 differnt users as well connected from this ip. This makes me think it is either the school itself possibly but the wierd time is unexplainged

Would also like to look into ip 97.105.19.61 if we had more time.

## 3. Are there students who, when active, hardly access the curriculum? If so, what information do you have about these students?



In [ ]:
# df.sort_values(by='count', ascending=False).head(2)

In [45]:
student_access_counts = df.groupby('user').size().sort_values().reset_index(name='access_count')

In [46]:
student_access_counts

user  access_count
0     592             1
1     652             1
2     649             1
3     918             1
4     879             1
..    ...           ...
976     1          7404
977   314          7783
978    53         12329
979    64         16347
980    11         17913

[981 rows x 2 columns]